text/3430-unix-socket-ancillary-data-v2.md

```diff
@@ -0,0 +1,896 @@
+    let mut ancillary = ancillary_buf.to_ancillary_data();
+
+    let mut buf = [0u8; 1];
+    MessageReceiver::new(stream, &mut buf)
+        .ancillary_data(&mut ancillary)
+        .recv()?;
+
+    let mut received_fds: Vec<_> = ancillary.received_fds().collect();
+    if received_fds.len() != 1 {
+        // TODO: error handling (std::io::Error if not enough FDs returned)
+        panic!("didn't receive enough FDs");
+    }
+    let received_fd = received_fds.pop().unwrap();
+    Ok(File::from(received_fd))
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Control messages
+
+Each control message is represented as a `ControlMessage` struct, which holds
+a reference to the underlying data.
+
+```rust
+struct ControlMessage<'a>;
+
+impl ControlMessage<'a> {
+    fn new(
+        cmsg_level: c_int,
+        cmsg_type: c_int,
+        data: &'a [u8],
+    ) -> ControlMessage<'a>;
+
+    // Opaque platform-specific integers; same as `struct cmsghdr` fields.
+    fn cmsg_level(&self) -> c_int;
+    fn cmsg_type(&self) -> c_int;
+
+    // The type-specific data of this control message.
+    fn data(&self) -> &[u8];
+
+    // Whether this control message is truncated, such as by being received
+    // into a too-short buffer.
+    fn truncated(&self) -> bool;
+}
+```
+
+A `&ControlMessages` represents a buffer full of control messages. It can be
+inspected and iterated over to obtain `ControlMessage` values.
+
+
+```rust
+struct ControlMessages;
+
+impl ControlMessages {
+    fn from_bytes(bytes: &[u8]) -> &ControlMessages;
+    fn as_bytes(&self) -> &[u8];
+    fn is_empty(&self) -> bool;
+    fn iter(&self) -> ControlMessagesIter<'_>;
+}
+
+impl<'a> IntoIterator for &'a ControlMessages {
+    type Item = ControlMessage<'a>;
+    type IntoIter = ControlMessagesIter<'a>;
+}
+
+struct ControlMessagesIter<'a>;
+
+impl<'a> Iterator for ControlMessagesIter<'a> {
+    type Item = ControlMessage<'a>;
+}
+
+impl ControlMessagesIter<'a> {
+    // For inspecting non-iterable fragment in truncated buffer.
+    fn into_bytes(self) -> &'a [u8];
+}
+```
+
+## Ancillary data
+
+### `struct AncillaryData`
+
+An `AncillaryData` is responsible for combining the serialized control messages
+with a notion of file descriptor ownership, ensuring that (1) borrowed FDs live
+long enough to be sent, and (2) received FDs aren't leaked.
+
+```rust
+struct AncillaryData<'a, 'fd>;
+
+impl Drop for AncillaryData;
+
+impl AncillaryData<'a, 'fd> {
+    fn new(
+        control_messages_buf: &'a mut [MaybeUninit<u8>],
+    ) -> AncillaryData<'a, 'fd>;
+
+    // returns initialized portion of `control_messages_buf`.
+    fn control_messages(&self) -> &ControlMessages;
+
+    // copy a control message into the ancillary data; error on out-of-capacity.
+    fn add_control_message(
+        &mut self,
+        control_message: impl Into<ControlMessage<'_>>,
```

Comment:
In this design, control messages are for structured viewing and appending to the `msg_control` buffer. Think of the `ControlMessage` struct as a Rustic equivalent to the `CMSG_*()` macros. They don't own file descriptors -- the ownership of FDs is managed by `AncillaryData`.

The purpose of the separate layers is to try to find the (rather small!) intersection between Rust's ownership semantics, the API idioms of early-1980s Unix, and a desire to avoid heap-allocating everything.

Thinking purely in terms of ownership, the logical signature of `sendmsg` / `recvmsg` might be notated something like this:

```rust
trait ControlMessage { ... }

sendmsg(
	&self,
	buf: &[u8],
	cmsg: Vec<Box<dyn ControlMessage>>,
        output_cmsg_capacity: usize,
	rights: Vec<BorrowedFd<'_>>,
) -> (io::Result<usize>, Vec<Box<dyn ControlMessage>>);

recvmsg(
	&self,
	buf: &mut [MaybeUninit<u8>],
	cmsg: Vec<Box<dyn ControlMessage>>,
        output_cmsg_capacity: usize,
) -> (io::Result<&mut [u8]>, Vec<Box<dyn ControlMessage>>, Vec<OwnedFd>);
```

In these signatures it's clear how the control messages are consumed and FDs are borrowed/produced. However, there's just no way such definitions could be used in Rust `std` -- besides the awful call sites, you'd have to heap-allocate everything *and* you'd need to bake in knowledge of all the various `SCM_*` types.

Supporting third-party libraries that want to add support for OS-specific `SCM_*` types means that we need to expose the actual raw byes of control messages in some reasonably accessible place -- hence `struct ControlMessage<'a>` providing a borrowed view of a `&'a ScmTimestampPktinfo` or whatever.

But if you're exposing raw cmsg bytes, they can't also be used for ownership of FDs without adding even more complexity to the API surface. I tried -- the original version of this had `ControlMessage<'fd>` and a `pub enum ScmRights { Borrowed(...),  Owned(..) }` -- but it just didn't work well.

Reply:
Hmm, but nothing prevents a user from adding `SCM_RIGHTS` manually using this function, and then the lifetime of the FDs isn't checked. So I still think this should have a lifetime for the resources the control message represents.

Even then, I think this function should be unsafe, as the control message contains instructions for the kernel. So who knows what it will do (the lifetime will still help prevent wrong usage though).

Reply:
> Hmm, but nothing prevents a user from adding `SCM_RIGHTS` manually using this function, and then the lifetime of the FDs isn't checked. So I still think this should have a lifetime for the resources the control message represents.

If we want to allow third-party code to implement arbitrary `SCM_*` control message types, then there's no way to prevent the user from manually encoding an `SCM_RIGHTS`.

> Even then, I think this function should be unsafe, as the control message contains instructions for the kernel. So who knows what it will do (the lifetime will still help prevent wrong usage though).

That's not what the `unsafe` keyword is for.

Reply:
It is, because we don't know what the kernel will do to our memory and open file descriptors because of those instructions.

Reply:
> If we want to allow third-party code to implement arbitrary `SCM_*` control message types, then there's no way to prevent the user from manually encoding an `SCM_RIGHTS`.

Yeah, but we can help them not make mistakes by adking for the correct lifetime.

Reply:
Are you sure that you can receive control messages with `sendmsg` and send them with `recvmsg`? The C API does technically allow it according to the pointer types, but I can't find it anywhere in documentation.

`nix` and `rustix` also don't support it, although of course that doesn't mean it's impossible:
https://docs.rs/nix/latest/nix/sys/socket/fn.sendmsg.html
https://docs.rs/nix/latest/nix/sys/socket/fn.recvmsg.html
https://docs.rs/rustix/latest/rustix/net/fn.sendmsg.html
https://docs.rs/rustix/latest/rustix/net/fn.recvmsg.html

Reply:
An example of control messages received from a `sendmsg()` is Linux's socket timestamps: https://www.kernel.org/doc/html/v6.6/networking/timestamping.html

Control messages passed *into* `recvmsg()` are sometimes used to configure socket options for that specific call only, though I forget the exact `SCM_*` type at the moment.

Reply:
That page says that timestamps are received with a `recvmsg()` call:

> 2 Data Interfaces
>
> Timestamps are read using the ancillary data feature of recvmsg(). See man 3 cmsg for details of this interface. The socket manual page (man 7 socket) describes how timestamps generated with SO_TIMESTAMP and SO_TIMESTAMPNS records can be retrieved.

The `SO_TIMESTAMP` and `SO_TIMESTAMPNS` options only apply to incoming packets:
> O_TIMESTAMP
>
>    Generates a timestamp for each incoming packet in (not necessarily monotonic) system time. <snip>
>
>    Same timestamping mechanism as SO_TIMESTAMP, but reports the timestamp as struct timespec in nsec resolution. <snip>


Also, in this interface you can enable timestamps through a CMSG only when sending packets:

> In addition to socket options, timestamp generation can be requested per write via cmsg, only for SOF_TIMESTAMPING_TX_* (see Section 1.3.1). Using this feature, applications can sample timestamps per sendmsg() without paying the overhead of enabling and disabling timestamps via setsockopt.


This all indicates that you can only send control messages with `sendmsg()` and you can only receive them with `recvmsg()`.

And if that's true, we can reflect that in the API, which can be simplified a little bit.

If you *can* send ancillary data with `recvmsg()` and you can receive it with `sendmsg()`, then I agree one struct for both is much easier. But it seems that you can't, and then splitting the API for the two separate use cases makes things a lot simpler.

Reply:
Hmm, interesting. I've written software in the past for a platform that modified the `sendmsg()` control messages buffer to report metadata about outgoing packet processing, but apparently it shouldn't have been doing that.

I'll want to do some more testing just to double-check, but if the read/write stages are strictly separated on all reasonable target platforms then that should let the new API be cleaner.

Thank you for pointing me in the right direction!